ros/src/twist_controller/twist_controller.py

```diff
@@ -1,14 +1,63 @@
-
+import rospy
+from pid import PID
+from yaw_controller import YawController
+from lowpass import LowPassFilter
 GAS_DENSITY = 2.858
-ONE_MPH = 0.44704
+FULL_BRAKE_SPEED = 0.1  # If target velocity is smaller, apply full brake
 
 
 class Controller(object):
-    def __init__(self, *args, **kwargs):
-        # TODO: Implement
-        pass
-
-    def control(self, *args, **kwargs):
-        # TODO: Change the arg, kwarg list to suit your needs
-        # Return throttle, brake, steer
-        return 1., 0., 0.
+    #Controller for vehicle brakes/ throttle / steering
+    def __init__(self, vehicle_mass, decel_limit, accel_limit, wheel_radius,
+                 wheel_base, steer_ratio, max_lat_accel, max_steer_angle,
+                 brake_deadband, fuel_capacity):
+
+        self.velocity_filter = LowPassFilter(3.5, 1.)
+        self.velocity_pid = PID(1., 0.001, 0.,
+                                mn=decel_limit, mx=0.25)
+        self.yaw_controller = YawController(wheel_base, steer_ratio, 1,
+                                            max_lat_accel, max_steer_angle)
+        self.wheel_radius = wheel_radius
+        self.brake_deadband = brake_deadband
+        self.total_mass = vehicle_mass + fuel_capacity * GAS_DENSITY
+        self.decel_limit = decel_limit
+        
+
+    def reset(self):
+        #Reset controller
+        self.velocity_pid.reset()
+
+    def control(self, twist, velocity, time_diff):
+        #Calc what contol needs to be applied based on desired vel and ang
+    	self.cmd_vel = twist.twist.linear.x
+        self.veh_vel = velocity.twist.linear.x
+        self.cmd_ang_vel = twist.twist.angular.z
+
+        velocity_cte = self.cmd_vel - self.veh_vel
+        cmd_acc = self.velocity_pid.step(velocity_cte, time_diff)
+        steer = self.yaw_controller.get_steering(self.cmd_vel,
+                                                 self.cmd_ang_vel,
+                                                 self.veh_vel)
+
+        # Apply low-pass filter to the linear acceleration
+        cmd_acc = self.velocity_filter.filt(cmd_acc)
+        # Keep full brake if target velocity is almost 0
+        if self.cmd_vel < FULL_BRAKE_SPEED:
+            throttle = 0.0
+            brake_torque = self.acceleration_to_torque(abs(self.decel_limit))
```

Comment:
The method `self.acceleration_to_torque` is not implemented which is the reason of the crash when the car finishes the lap.
However it seems that everything works fine if we always use the `else` branch of the test.

Reply:
OK, i had a function to convert the accel/decel to torque but I ended up removing it, I will adjust the brake torque in the if statement. What is the best way to change this? New pull request?

Reply:
I believe, you can amend the commit cff49d6 and push again.

Reply:
It is recommended to not amend commits that have been already pushed. It is better to do a new commit on top of the branch and then push the branch again.